src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs

```diff
@@ -0,0 +1,1188 @@
+    // - Lookup by key
+    // - Indexing by position
+    // - Adding
+    // - Inserting
+    // - Removing
+    //
+    // There's not a good way to achieve all of those, e.g.
+    // - A map for lookups with an array list achieves O(1) lookups, indexing, and adding, but O(N) insert and removal.
+    // - A map for lookups with a linked list achieves O(1) lookups, adding, removal, and insert, but O(N) indexing.
+    //
+    // There are also layout and memory consumption tradeoffs. For example, a map to nodes containing keys and values
+    // means lots of indirections as part of enumerating. Alternatively, the keys and values can be duplicated in both
+    // a map and a list, leading to larger memory consumption, but optimizing for speed of data access. Or the keys
+    // and values can be stored in the map with only the key stored in the list.
+    //
+    // This implementation currently employs the simple strategy of using both a dictionary and a list, with the
+    // dictionary as the source of truth for the key/value pairs, and the list storing just the keys in order. This
+    // provides O(1) lookups, adding, and indexing, with O(N) insert and removal. Keys are duplicated in memory,
+    // but lookups are optimized to be simple dictionary accesses. Enumeration is O(N), and involves enumerating
+    // the list for order and performing a lookup on each element to get its value. This is the same approach taken
+    // by the non-generic OrderedDictionary and thus keeps algorithmic complexity consistent for someone upgrading
+    // from the non-generic to generic types. It's also important for consumption via the interfaces, in particular
+    // I{ReadOnly}List<T>, where it's common to iterate through a list with an indexer, and if indexing were O(N)
+    // instead of O(1), it would turn such loops into O(N^2) instead of O(N).
+    //
+    // Currently the implementation is optimized for simplicity and correctness, choosing to wrap a Dictionary<>
+    // and a List<> rather than implementing a custom data structure. They could be flattened to partially
+    // deduped in the future if the extra overhead is deemed prohibitive.
+
+    /// <summary>
+    /// Represents a collection of key/value pairs that are accessible by the key or index.
+    /// </summary>
+    /// <typeparam name="TKey">The type of the keys in the dictionary.</typeparam>
+    /// <typeparam name="TValue">The type of the values in the dictionary.</typeparam>
+    [DebuggerTypeProxy(typeof(IDictionaryDebugView<,>))]
+    [DebuggerDisplay("Count = {Count}")]
+    public class OrderedDictionary<TKey, TValue> :
+        IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>, IDictionary,
+        IList<KeyValuePair<TKey, TValue>>, IReadOnlyList<KeyValuePair<TKey, TValue>>, IList
+        where TKey : notnull
+    {
+        /// <summary>Store for the key/value pairs in the dictionary.</summary>
+        private readonly Dictionary<TKey, TValue> _dictionary;
```

Comment:
Is there a reason we're not using a dedicated data store like in my existing implementation? https://github.com/dotnet/corefxlab/blob/archive/src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

Reply:
Because it adds a lot of complexity, especially to mirror all the optimizations in `Dictionary<>`, e.g. that existing implementation doesn't special-case strings. If we want to optimize it later based on demonstrated need that it's worth the complexity, we can.